thewalrus/quantum.py

```diff
@@ -1040,27 +1043,48 @@ def fock_tensor(S, alpha, cutoff, choi_r=np.arcsinh(1.0), check_symplectic=True,
 
     # And that S and alpha have compatible dimensions
     m, _ = S.shape
-    if m // 2 != len(alpha):
-        raise ValueError("The matrix S and the vector alpha do not have compatible dimensions")
-
-    # Construct the covariance matrix of l two-mode squeezed vacua pairing modes i and i+l
     l = m // 2
-    ch = np.cosh(choi_r) * np.identity(l)
-    sh = np.sinh(choi_r) * np.identity(l)
-    zh = np.zeros([l, l])
-    Schoi = np.block([[ch, sh, zh, zh], [sh, ch, zh, zh], [zh, zh, ch, -sh], [zh, zh, -sh, ch]])
-    # And then its Choi expanded symplectic
-    S_exp = expand(S, list(range(l)), 2 * l) @ Schoi
-    # And this is the corresponding covariance matrix
-    cov = S_exp @ S_exp.T
-    alphat = np.array(list(alpha) + ([0] * l))
-    x = 2 * alphat.real
-    p = 2 * alphat.imag
-    mu = np.concatenate([x, p])
-
-    tensor = state_vector(
-        mu, cov, normalize=False, cutoff=cutoff, hbar=2, check_purity=False, choi_r=choi_r
-    )
+    if l != len(alpha):
+        raise ValueError(
+            "The matrix S and the vector alpha do not have compatible dimensions"
+        )
+    # Check if S corresponds to an interferometer, if so use optimized routines
+    if np.allclose(S @ S.T, np.identity(m)) and np.allclose(alpha, 0):
+        reU = S[:l, :l]
+        imU = S[:l, l:]
+        if np.allclose(imU, 0):

```

Comment:
do you want to have the tolerance here modified as a function argument?

Reply:
I think it would be a good idea to have an `atol` and `rtol` arguments be passed by the user and take a default value to be used in all the `np.allclose`. It would make this function have a very long signature but I think that is ok.